### ISSUE TITLE
[Bug]: LKE Cluster Pool count must be optional to match API and avoid recreates

### ISSUE THREAD

**Issue by displague** [linode/terraform-provider-linode#1076]
### Description

When using the autoscaler, the pool size (`count`) should be optional.

Pool count is currently required in TF. The API requires this field on creation. The API does not require node pool `count` on updates.
* https://www.linode.com/docs/api/linode-kubernetes-engine-lke/#kubernetes-cluster-create
* https://www.linode.com/docs/api/linode-kubernetes-engine-lke/#node-pool-create
* https://www.linode.com/docs/api/linode-kubernetes-engine-lke/#node-pool-update

When updating, while the count has been modified by the autoscaler, the cluster will attempt to reset the pool count. This is an unnecessary change and forces nodes to be recycled. (Seen in #782 and comments below)

If the pool count parameter were made optional in the presence of an autoscaler setting, a minimum size could be asserted for creates (pool min size) and on updates the count field could be omitted in API requests when not defined by TF configuration.

```
# module.lke.linode_lke_cluster.lke will be updated in-place
  ~ resource "linode_lke_cluster" "lke" {
        id            = "46246"
        tags          = [
            "lke",
            "terraform",
        ]
        # (7 unchanged attributes hidden)

      ~ pool {
          ~ count = 0 -> 3
            id    = 0
          + type  = "g6-standard-2"
            # (1 unchanged attribute hidden)

          + autoscaler {
              + max = 3
              + min = 1
            }
        }
      - pool {
          - count = 1 -> null
          - id    = 74170 -> null
          - nodes = [
              - {
                  - id          = "74170-3f9682960000"
                  - instance_id = 50828728
                  - status      = "ready"
                },
            ] -> null
          - type  = "g6-standard-2" -> null

          - autoscaler {
              - max = 3 -> null
              - min = 1 -> null
            }
        }

        # (1 unchanged block hidden)
    }
```

### New or Affected Terraform Resources

linode_cluster_pool

### Potential Terraform Configuration

```hcl
resource "linode_lke_cluster" "lke" {
  label       = "lke.tf"
  k8s_version = var.k8s_version
  region      = "us-central"
  tags        = ["lke", "terraform"]

  pool {
    type  = "g6-standard-2"
    # count is omitted
    autoscaler {
      min = 1
      max = 3
    }
  }
}
```

**Comment by zliang-akamai** [comment]
Thanks @displague!
We will investigate it some time next week.

**Comment by aslepcev** [comment]
Hey @displague , you could use the lifecycle meta argument to tell Terraform to ignore the LKE pool VM count in case autoscale changes it. 
Here's an example code. 

`resource "linode_lke_cluster" "akamai-sync" {
    label       = "Akamai-Data-Sync"
    k8s_version = "1.27"
    region      = "us-lax"
    tags        = ["foo"]
    
    control_plane {
      high_availability = true
    }
    
    pool {
        type  = "g6-standard-2"
        count = 3

      autoscaler {
          min = 1
          max = 10
        }        
    }

  # Prevent the count field from overriding autoscaler-created nodes
  lifecycle {
    ignore_changes = [
      pool[0].count
    ]
  }
}`

After running terraform apply, I manually scaled the pool to 5 nodes and reran terrafom apply. Pool count remained at 5 and this is the output of the terraform plan: 

`Terraform will perform the following actions:

  # linode_lke_cluster.akamai-sync will be updated in-place
  ~ resource "linode_lke_cluster" "akamai-sync" {
        id            = "137431"
        tags          = [
            "app:datasync",
        ]
        # (7 unchanged attributes hidden)

      ~ pool {
            id    = 0
          + type  = "g6-standard-2"
            # (2 unchanged attributes hidden)

          + autoscaler {
              + max = 10
              + min = 1
            }
        }
      - pool {
          - count = 5 -> null
          - id    = 202422 -> null
          - nodes = [
              - {
                  - id          = "202422-02db5c6f0000"
                  - instance_id = 51085556
                  - status      = "ready"
                },
              - {
                  - id          = "202422-103d2c780000"
                  - instance_id = 51085525
                  - status      = "ready"
                },
              - {
                  - id          = "202422-167aa3ce0000"
                  - instance_id = 51085464
                  - status      = "ready"
                },
              - {
                  - id          = "202422-50b9bc020000"
                  - instance_id = 51085459
                  - status      = "ready"
                },
              - {
                  - id          = "202422-650faa6f0000"
                  - instance_id = 51085463
                  - status      = "ready"
                },
            ] -> null
          - type  = "g6-standard-2" -> null

          - autoscaler {
              - max = 10 -> null
              - min = 1 -> null
            }
        }

        # (1 unchanged block hidden)
    }`

Regards, Alex.

**Comment by displague** [comment]
That's a reasonable and practical workaround, @aslepcev. Thanks. I'll use that in my module. 

The Terraform model doesn't match the API model. This could be addressed in the provider in an intuitive and non-breaking way that doesn't require life cycle overrides. I'd like to keep this issue open for consideration, but I'll understand if you want to close it for organizational purposes.

**Comment by dineshsaathe** [comment]
I am facing the same issue even with the "ignore_changes". After applying it shows modification complete but on the very next plan it shows the exact same output.


` ~ resource "linode_lke_cluster" "cluster" {
        id            = "12345"
        tags          = [
            "prod",
        ]
        # (7 unchanged attributes hidden)

      ~ pool {
            id    = 0
          + type  = "g6-standard-4"
            # (2 unchanged attributes hidden)

          + autoscaler {
              + max = 10
              + min = 1
            }
        }
      - pool {
          - count = 3 -> null
          - id    = 197627 -> null
          - nodes = [
              - {
                  - id          = "13458-147804480000"
                  - instance_id = 50679426
                  - status      = "ready"
                },
              - {
                  - id          = "13458-325823200000"
                  - instance_id = 50800684
                  - status      = "ready"
                },
              - {
                  - id          = "197627-57469dbb0000"
                  - instance_id = 51900214
                  - status      = "ready"
                },
            ] -> null
          - type  = "g6-standard-4" -> null

          - autoscaler {
              - max = 10 -> null
              - min = 1 -> null
            }
        }

        # (3 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
`

**Comment by displague** [comment]
@dineshsaathe since you have multiple pools, did you ignore_changes for both:

```hcl
  lifecycle {
    ignore_changes = [
      pool[0].count,
      pool[1].count
    ]
  }
```

**Comment by displague** [comment]
@aslepcev @lgarber-akamai the workaround doesn't work, as @dineshsaathe and #782 report.


```tf
resource "linode_lke_cluster" "lke" {
  label       = "lke.tf"
  k8s_version = var.k8s_version
  region      = "us-central"
  tags        = ["lke", "terraform"]

  pool {
    type  = "g6-standard-2"
    count = var.min_count

    autoscaler {
      min = var.min_count
      max = var.max_count
    }
  }

  lifecycle {
    ignore_changes = [
      pool[0].count,
    ]
  }
}
```

```
  # module.lke.linode_lke_cluster.lke will be updated in-place
  ~ resource "linode_lke_cluster" "lke" {
        id            = "46246"
        tags          = [
            "lke",
            "terraform",
        ]
        # (7 unchanged attributes hidden)

      ~ pool {
            id    = 0
          + type  = "g6-standard-2"
            # (2 unchanged attributes hidden)

          + autoscaler {
              + max = 3
              + min = 1
            }
        }
      - pool {
          - count = 3 -> null
          - id    = 74170 -> null
          - nodes = [
              - {
                  - id          = "74170-18d442530000"
                  - instance_id = 53000971
                  - status      = "ready"
                },
              - {
                  - id          = "74170-1d7b62c40000"
                  - instance_id = 53000973
                  - status      = "ready"
                },
              - {
                  - id          = "74170-302994550000"
                  - instance_id = 50832943
                  - status      = "ready"
                },
            ] -> null
          - type  = "g6-standard-2" -> null

          - autoscaler {
              - max = 3 -> null
              - min = 1 -> null
            }
        }

        # (1 unchanged block hidden)
    }
```

**Comment by lseelenbinder** [comment]
I can confirm @displague's issue. Ignoring doesn't work for me, either. This really makes it difficult to work with LKE clusters in TF. Any timeline on a fix for this? (It should honestly be a #bug not a #feature request IMO. 😄)

Thanks!

**Comment by yec-akamai** [comment]
Thank you all for reporting the issue!

We've addressed it in #1260 and released it in [v.2.14.0](https://github.com/linode/terraform-provider-linode/releases/tag/v2.14.0). Please let us how it works and feel free to reopen it if any issue comes up.
